src/java.base/share/classes/java/net/URLEncoder.java

```diff
@@ -122,22 +124,45 @@ public class URLEncoder {
          *
          */
 
-        var bitSet = new BitSet(128);
-        bitSet.set('a', 'z' + 1);
-        bitSet.set('A', 'Z' + 1);
-        bitSet.set('0', '9' + 1);
-        bitSet.set(' '); /* encoding a space to a + is done
-                                    * in the encode() method */
-        bitSet.set('-');
-        bitSet.set('_');
-        bitSet.set('.');
-        bitSet.set('*');
+        long flag0 = 0;
+        flag0 |= 1L << ' '; // ASCII 32
+        flag0 |= 1L << '*'; // ASCII 42
+        flag0 |= 1L << '-'; // ASCII 25
+        flag0 |= 1L << '.'; // ASCII 46
+
+        // ASCII 48 - 57
+        for (int i = '0'; i <= '9'; ++i) {
+            flag0 |= 1L << i;
+        }
+        DONT_NEED_ENCODING_FLAGS_0 = flag0;
 
-        DONT_NEED_ENCODING = ImmutableBitSetPredicate.of(bitSet);
+        long flags1 = 0;
+        // ASCII 65 - 90
+        for (int i = 'A'; i <= 'Z'; ++i) {
+            flags1 |= 1L << (i - 64);
+        }
+        flags1 |= 1L << ('_' - 64); // ASCII 95
+        // ASCII 97 - 122
+        for (int i = 'a'; i <= 'z'; ++i) {
+            flags1 |= 1L << (i - 64);
+        }
+        DONT_NEED_ENCODING_FLAGS_1 = flags1;
 
         DEFAULT_ENCODING_NAME = StaticProperty.fileEncoding();
     }
 
+    /**
+     * dotNeedEncoding
+     */
+    private static boolean dontNeedEncoding(int c) {
+        int prefix = (c >>> 6);
+        if (prefix > 1) {
+            return false;
+        }
+        long flags = prefix == 0 ? DONT_NEED_ENCODING_FLAGS_0 : DONT_NEED_ENCODING_FLAGS_1;
+        return (flags & (1L << (c & 0x3f))) != 0;
```

Comment:
```suggestion
        return (flags & (1L << c)) != 0;
```
`c & 0x3f` is redundant; you can test in `jshell` that `1L << 65` is same as `1L << 1`, both `2`